modules/ingest-common/src/test/java/org/elasticsearch/ingest/common/RenameProcessorTests.java

```diff
@@ -171,4 +171,27 @@ public Object remove(Object key) {
             assertThat(ingestDocument.getSourceAndMetadata().containsKey("new_field"), equalTo(false));
         }
     }
+
+    public void testRenameLeafIntoBranch() throws Exception {
+        Map<String, Object> source = new HashMap<>();
+        source.put("foo", "bar");
+        IngestDocument ingestDocument = new IngestDocument(source, Collections.emptyMap());
+        Processor processor1 = new RenameProcessor(randomAsciiOfLength(10), "foo", "foo.bar");
+        processor1.execute(ingestDocument);
+        assertThat(ingestDocument.getFieldValue("foo", Map.class), equalTo(Collections.singletonMap("bar", "bar")));
+        assertThat(ingestDocument.getFieldValue("foo.bar", String.class), equalTo("bar"));
+
+        Processor processor2 = new RenameProcessor(randomAsciiOfLength(10), "foo.bar", "foo.bar.baz");
+        processor2.execute(ingestDocument);
+        assertThat(ingestDocument.getFieldValue("foo", Map.class), equalTo(Collections.singletonMap("bar",
+                Collections.singletonMap("baz", "bar"))));
+        assertThat(ingestDocument.getFieldValue("foo.bar", Map.class), equalTo(Collections.singletonMap("baz", "bar")));
+        assertThat(ingestDocument.getFieldValue("foo.bar.baz", String.class), equalTo("bar"));
+
+        // for fun lets try to restore it (which don't allow today)
```

Comment:
this may get confusing since the feature will be allowed `today`, where `today` is some time in the future that someone will read this. maybe we can reference the PR here, and use more past-tense terms like `previously`.